[processor/tailsampling] Expose Sampling Policy interfaces

csmarchbanks opened the pull request:
#### Description

Refactor the tail sampling processor so that the interfaces that define trace data and policy decisions are available publically. This is the first step in allowing extensions to be added to the tail sampling processor to allow custom logic that is not possible/easy to do with OTTL such as pulling information from other sources.

#### Link to tracking issue
Part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31582

#### Testing
No behavior changed as this is purely a refactor. All current tests need to continue to pass.

csmarchbanks committed:
Expose Sampling Policy interfaces

Refactor the tail sampling processor so that the interfaces that define
trace data and policy decisions are available publically. This is the
first step in allowing extensions to be added to the tail sampling
processor to allow custom logic that is not possible/easy to do with
OTTL such as pulling information from other sources.

axw commented on processor/tailsamplingprocessor/and_helper.go:
```diff
@@ -7,10 +7,11 @@ import (
 
-func getNewAndPolicy(settings component.TelemetrySettings, config *AndCfg) (sampling.PolicyEvaluator, error) {
-	subPolicyEvaluators := make([]sampling.PolicyEvaluator, len(config.SubPolicyCfg))
+func getNewAndPolicy(settings component.TelemetrySettings, config *AndCfg) (samplingpolicy.PolicyEvaluator, error) {
```
Now that PolicyEvaluator is now in a package called samplingpolicy, I think it would be natural to rename it to Evaluator. That also aligns nicely with the method name Evaluate.

axw reviewed (commented):
Looks good, just one suggestion to rename the interface

csmarchbanks committed:
Rename PolicyEvaluator to Evaluator

Now that the interface has its own samplingpolicy package it makes sense
nto to repeat policy.

axw reviewed (approved):
LGTM!

csmarchbanks committed:
More renames I missed when running tests the first time...

csmarchbanks committed:
Format code

atoulme commented:
Please resolve conflicts.

csmarchbanks commented:
Rebased, though for some reason CI is mostly sitting in a pending state now.
